### ISSUE TITLE
clang 17.0.2: No "unreachable code" warning if source file included in another source file

### ISSUE THREAD

**Issue by pkl97** [llvm/llvm-project#71046]
When compiling unreachable_warning.cpp (with the content below) using option -Wunreachable-code-break, a warning is triggered as expected:
```c++
void f(int i)
{
    switch (i)
    {
        case 3: return; break;
    }
}
```
The warning emitted:
```console
[user@host]$ clang++ -c -Wunreachable-code-break unreachable_warning.cpp
unreachable_warning.cpp:5:25: warning: 'break' will never be executed [-Wunreachable-code-break]
        case 3: return; break;
                        ^~~~~
1 warning generated.
```

However if unreachable_warning.cpp is wrapped in another source file (as is frequently done in "unity builds" or "jumbo builds") then the warning is no longer emitted:
```
[user@host]$ cat unity_build.cpp
#include "unreachable_warning.cpp"

[user@host]$ clang++ -c -Wunreachable-code-break unity_build.cpp
[user@host]$
```

**Comment by shafik** [comment]
This looks purposeful:

https://github.com/llvm/llvm-project/blob/fee2953f23bd8a8a71e574e6a8db08033778d3a4/clang/lib/Sema/AnalysisBasedWarnings.cpp#L125-L134

CC @AaronBallman

**Comment by AaronBallman** [comment]
Correct, this is behaving by design and I don't think we intend to change the behavior here due to risk of significant increase to compile times.

**Comment by pkl97** [comment]
Thanks @shafik for the thorough investigation!

@AaronBallman:
Is there any chance to relax this feature a bit? Compile time increase should only be an issue if dealing with actual header files (files with extensions like .h, .hh, .hpp, .hxx) that are included from many different source files.
If the included file is a source file however, the inclusion only occurs once across the whole project, in a single translation unit. So there would be no overhead, except applying the warning to a source file for which it was intended in the first place.
The code could check for an extension like .c, .cc, .cpp, .cxx and only apply the warning in those cases. This should be a rather safe approach.

**Comment by AaronBallman** [comment]
> Thanks @shafik for the thorough investigation!
> 
> @AaronBallman: Is there any chance to relax this feature a bit? Compile time increase should only be an issue if dealing with actual header files (files with extensions like .h, .hh, .hpp, .hxx) that are included from many different source files. If the included file is a source file however, the inclusion only occurs once across the whole project, in a single translation unit. So there would be no overhead, except applying the warning to a source file for which it was intended in the first place. The code could check for an extension like .c, .cc, .cpp, .cxx and only apply the warning in those cases. This should be a rather safe approach.

There's always a chance! :-) I generally agree with your logic, but I think this leads to a bit of a game of whack-a-mole with compiler or language features that don't work so well in unity builds. I'm wondering if a better approach would be to design a flag to enable "unity build mode" that configures the compiler to behave in a nicer manner for this use case. For example, this could also allow us to diagnose linkage issues (same identifier in anonymous namespaces that suddenly link together), macro leakage, and other unity-build-specific issues. The downside is: it's also a much bigger feature request than what you're after.

**Comment by dwblaikie** [comment]
(& at the point we're talking about unity build support - I tend to go further and wonder what problems we have in other infrastructure (LTO, ThinLTO) that could be improved to address the use case for unity builds in a more scalable manner (& one more consistent with other users/use cases we have))

**Comment by vszakats** [comment]
With unity, this particular issue is the first downside we hit so far.

IMO LTO and unity/umbo target different problems. unity is useful to make the build (much) faster, while LTO is making the result run faster with usually slower (but certainly not much faster) build times. (since link time, it also involves dealing with dependencies and potential issues with them.) Disk footprint also has different properties.

It's also not just unity mode. There can be other legitimate reasons to include a `.c` source into another one. libssh2 uses it to include the selected crypto backend code without resorting to build/make-specific tricks.

The suggested method to make disctinction between included headers vs. source via suffix sounds nice and would solve both unity and other cases without the complications of new compiler option.

**Comment by dwblaikie** [comment]
> With unity, this particular issue is the first downside we hit so far.
> 
> IMO LTO and unity/umbo target different problems. unity is useful to make the build (much) faster, while LTO is making the result run faster with usually slower (but certainly not much faster) build times. (since link time, it also involves dealing with dependencies and potential issues with them.) Disk footprint also has different properties.

Ah, fair enough - my mistake. Header modules might be more relevant, then.

> It's also not just unity mode. There can be other legitimate reasons to include a `.c` source into another one. libssh2 [uses it](https://github.com/libssh2/libssh2/blob/d01680ca09cdc49ebb983b62cf5ab1db923cb218/src/crypto.c) to include the selected crypto backend code without resorting to build/make-specific tricks.
> 
> The [suggested method](https://github.com/llvm/llvm-project/issues/71046#issuecomment-1798274347) to make disctinction between included headers vs. source via suffix sounds nice and would solve both unity and other cases without the complications of a new compiler option.

Yeah - I suspect there's an underlying issue that this warning just isn't very good/robust/fully developed. So I'd mostly just discourage people from using it for now - not sure how I feel about making an extension based filter... I guess I wouldn't outright say no, and it's probably more @AaronBallman's call, but I'm not super keen, let's say.

**Comment by vszakats** [comment]
Thanks for your insights.

I feel your sentiments. It'd be a pragmatic solution, or rather, workaround.

Another way may be to suppress this check for system headers (via `-isystem`), as done for most other warnings. By all likelyhood it was already considered, and would need to address robustness/perf first.

Perhaps the best is indeed disabling it for projects supporting unity.
